components/script/dom/htmlimageelement.rs

```diff
@@ -271,101 +264,319 @@ impl HTMLImageElement {
+                // Step 11, substep 5
+                let img = self.img.root();
+                img.current_request.borrow_mut().source_url = Some(self.src.into());
+                img.upcast::<EventTarget>().fire_event(atom!("error"));
+                img.upcast::<EventTarget>().fire_event(atom!("loadend"));
+            }
+        }
+        let runnable = box SetUrlToSelectedTask {
+            img: Trusted::new(self),
+            src: src.into(),
+        };
+        let document = document_from_node(self);
+        let window = document.window();
+        let task = window.dom_manipulation_task_source();
+        let _ = task.queue(runnable, window.upcast());
+    }
+
+    /// Step 12 of html.spec.whatwg.org/multipage/#update-the-image-data
+    fn fetch_image(&self, url: &ServoUrl, src: &DOMString) {
+        let mut current_request = self.current_request.borrow_mut();
+        let mut pending_request = self.pending_request.borrow_mut();
+        if let Some(pending_url) = pending_request.parsed_url.clone() {
+            // Step 12.1
+            if pending_url == *url {
+                return
+            }
+            // Step 12.2, 12.3 abort pending request
+            pending_request.image = None;
+            pending_request.parsed_url = None;
+            LoadBlocker::terminate(&mut pending_request.blocker);
+        }
+        // step 12.4, create a new "image_request"
+        let mut image_request = match (current_request.parsed_url.clone(), current_request.state) {
+            (Some(parsed_url), State::PartiallyAvailable) => {
+                // Step 12.2
+                if parsed_url == *url {
+                    // TODO: queue a task to restart animation, if restart-animation is set
+                    return
+                }
+                pending_request
+            },
+            (_, State::Broken) | (_, State::Unavailable) => {
+                // Step 12.5
+                LoadBlocker::terminate(&mut current_request.blocker);
+                current_request
+            },
+            (_, _) => {
+                // step 12.6
+                pending_request
+            },
+        };
+        image_request.parsed_url = Some(url.clone());
+        image_request.source_url = Some(src.clone());
+        image_request.image = None;
+        image_request.metadata = None;
+        let document = document_from_node(self);
+        image_request.blocker = Some(LoadBlocker::new(&*document, LoadType::Image(url.clone())));
+        // Step 12.10 associate this instance of algorithm with "image_request"
+        // we do this via the url.
+        self.update_image_with_url(&url);
+    }
+
+    /// Step 8-12 of html.spec.whatwg.org/multipage/#update-the-image-data
+    fn update_the_image_data_sync_steps(&self) {
+        let document = document_from_node(self);
+        // Step 8
+        let image_source = self.select_image_source();
```

Comment:
On this one I have a question: have you considered that the spec mentions "selecting a source" as part of the sync steps, once a so-called 'stable state' has been reached, and that the sync steps will run only once, even if `update_the_image_data` was called several times. Since `update_the_image_data` should be called if the source changes, perhaps when we reach this point we in fact will always get the "latest" source?